### PR TITLE
core(installable-manifest): expose app manifest url

### PR DESCRIPTION
**Pull request by adrianaixba**
for #9683

exposes the app manifest url in the installable manifest audit

### CODE REVIEW

File: lighthouse-core/gather/gather-runner.js [hunk 1]
```diff
@@ -626,12 +628,15 @@ class GatherRunner {
...
    * @param {LH.Gatherer.PassContext} passContext
-   * @return {Promise<LH.Artifacts.Manifest|null>}
+   * @return {Promise<{parsedManifest: LH.Artifacts.Manifest|null, manifestUrl: string|null} | null>}
    */
   static async getWebAppManifest(passContext) {
     const response = await passContext.driver.getAppManifest();
     if (!response) return null;
-    return manifestParser(response.data, response.url, passContext.url);
+    const parsedManifest = manifestParser(response.data, response.url, passContext.url);
+    const manifestUrl = response.url;
```
Comment (paulirish) [inline lighthouse-core/gather/gather-runner.js:637]: we talked in voice about this.. should manifestUrl be part of the parser's return value?  I said no, but... let's reconsider. ... it could. and it kinda makes sense. just spoke with @brendankenny and we both feel fairly ambivalent about it.

It keeps things a tad cleaner to put it in the manifestParser return, so let's do that. (well, as it turns out... you [already did](https://github.com/GoogleChrome/lighthouse/pull/11330/files#diff-af5212b94825d538f81ee91d73c6e326))

that actually means you can revert allllll the changes in gather-runner

anddddddddddd i think everything will just still end up working fine.

File: lighthouse-core/audits/installable-manifest.js [hunk 2]
```diff
@@ -89,17 +89,20 @@ class InstallableManifest extends MultiCheckAudit {
...
-   * @return {Promise<{failures: Array<string>, manifestValues: LH.Artifacts.ManifestValues}>}
+   * @return {Promise<{failures: Array<string>, manifestValues: LH.Artifacts.ManifestValues, manifestUrl: string | undefined}>}
    */
   static async audit_(artifacts, context) {
     const manifestValues = await ManifestValues.request(artifacts, context);
     const manifestFailures = InstallableManifest.assessManifest(manifestValues);
 
+    const manifestUrl = artifacts.WebAppManifest ? artifacts.WebAppManifest.manifestUrl : undefined;
+
     return {
```
Comment (paulirish) [inline lighthouse-core/audits/installable-manifest.js:100]: side note how do we have an audit that doesn't have a `score`.. so confused

Comment (patrickhulce) [reply lighthouse-core/audits/installable-manifest.js:100]: `extends MultiCheckAudit` and that `audit_` life, def confusing though

Comment (paulirish) [reply lighthouse-core/audits/installable-manifest.js:100]: OHHHH MultiCheckAudit. of course.

File: lighthouse-core/lib/manifest-parser.js [hunk 3]
```diff
@@ -505,6 +505,7 @@ function parse(string, manifestUrl, documentUrl) {
     raw: string,
     value: manifest,
     warning: manifestUrlWarning,
+    manifestUrl: manifestUrl,
```
Comment (connorjclark) [inline lighthouse-core/lib/manifest-parser.js:508]: I think `url` would be a better name here, considering the natural way to name a variable holding this object is `manifest` (gather-runner actually uses `parsedManifest` which is also cool), and `manifest.manifestUrl` is redundant.

I think `manifest.url` would not lead to any confusion about what the url is referring to.

File: lighthouse-core/gather/gather-runner.js [hunk 4]
```diff
@@ -584,7 +584,9 @@ class GatherRunner {
     }
 
     // Fetch the manifest, if it exists.
-    baseArtifacts.WebAppManifest = await GatherRunner.getWebAppManifest(passContext);
+    const manifestResult = await GatherRunner.getWebAppManifest(passContext);
+    // manifestResult could be null
```
Comment (connorjclark) [inline lighthouse-core/gather/gather-runner.js:588]: nit: we follow this comment style guide:

```js
// Start with capital letter. End with punctuation.
```

Comment (connorjclark) [reply lighthouse-core/gather/gather-runner.js:588]: (oh, but it does seem like we can revert these few lines?)

File: lighthouse-core/gather/gather-runner.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/gather/gather-runner.js:637]: yea it seems like this whole file could be reverted and the url from the parser lib will be consumable in `installable-manifest`

File: lighthouse-core/audits/installable-manifest.js [hunk 5]
```diff
@@ -89,17 +89,20 @@ class InstallableManifest extends MultiCheckAudit {
...
    * @param {LH.Artifacts} artifacts
    * @param {LH.Audit.Context} context
-   * @return {Promise<{failures: Array<string>, manifestValues: LH.Artifacts.ManifestValues}>}
+   * @return {Promise<{failures: Array<string>, manifestValues: LH.Artifacts.ManifestValues, manifestUrl: string | undefined}>}
    */
   static async audit_(artifacts, context) {
     const manifestValues = await ManifestValues.request(artifacts, context);
     const manifestFailures = InstallableManifest.assessManifest(manifestValues);
 
+    const manifestUrl = artifacts.WebAppManifest ? artifacts.WebAppManifest.manifestUrl : undefined;
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/audits/installable-manifest.js:98]: yea `null` is preferable for "missing", especially when the object may be JSON serialized (properties set to `undefined` are dropped in JSON)

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 6]
```diff
@@ -1798,8 +1798,9 @@ describe('GatherRunner', function() {
         .mockResponse('Page.getInstallabilityErrors', {installabilityErrors: []});
 
       const result = await GatherRunner.getWebAppManifest(passContext);
-      expect(result).toHaveProperty('raw', JSON.stringify(manifest));
-      expect(result && result.value).toMatchObject({
+      const parsedManifest = result ? result.parsedManifest : null;
+      expect(parsedManifest).toHaveProperty('raw', JSON.stringify(manifest));
```
Comment (connorjclark) [inline lighthouse-core/test/gather/gather-runner-test.js:1802]: this can be reverted to after you revert the changes in `gather-runner`, but I think an assertion here for `result.url` would be good.

File: lighthouse-core/lib/manifest-parser.js (hunk 3 above)
Comment (adrianaixba) [reply lighthouse-core/lib/manifest-parser.js:508]: Do you think we should do this for the audit items value too? details.items[0].url vs details.items[0].manifestUrl. manifestUrl does add some clarity, although not sure if its necessary given that the audit refers to the manifest.

Comment (connorjclark) [reply lighthouse-core/lib/manifest-parser.js:508]: `url` would be great there too

Comment (brendankenny) [reply lighthouse-core/lib/manifest-parser.js]: There is some pretty good self documenting value for keeping it `manifestUrl` when it's deep in `installableManifest.details.items[0]`, though, since we don't publish docs for specific audit's details.

Note also it's a [term from the spec](https://www.w3.org/TR/appmanifest/#processing), to differentiate from "document URL" and `start_url`.

**Comment by connorjclark** [comment]
fyi if you put "fixed by #blah" in the first post, when the PR merges it auto-closes the associated issue
